templates/commands/upgrade/upgrade.go

```diff
@@ -140,29 +140,32 @@ func (c *Command) Run(ctx context.Context, args []string) error {
 		return err //nolint:wrapcheck
 	}
 
-	if r.AlreadyUpToDate {
+	switch r.Type {
+	case upgrade.AlreadyUpToDate:
 		fmt.Fprintf(c.Stdout(), "Already up to date with latest template version\n")
 		return nil
-	}
-
-	if len(r.Conflicts) == 0 {
+	case upgrade.Success:
 		fmt.Fprintf(c.Stdout(), "Upgrade complete with no conflicts\n")
 		return nil
-	}
-
-	// TODO(upgrade):
-	//  - suggest diff / meld / vim commands?
-	fmt.Fprint(c.Stdout(), mergeInstructions+"\n\n--\n")
-	for _, cf := range r.Conflicts {
-		fmt.Fprintf(c.Stdout(), "file: %s\n", cf.Path)
-		fmt.Fprintf(c.Stdout(), "conflict type: %s\n", cf.Action)
-		if cf.OursPath != "" {
-			fmt.Fprintf(c.Stdout(), "our file was renamed to: %s\n", cf.OursPath)
-		}
-		if cf.IncomingTemplatePath != "" {
-			fmt.Fprintf(c.Stdout(), "incoming file: %s\n", cf.IncomingTemplatePath)
+	case upgrade.MergeConflict:
+		// TODO(upgrade):
+		//  - suggest diff / meld / vim commands?
+		fmt.Fprint(c.Stdout(), mergeInstructions+"\n\n--\n")
+		for _, cf := range r.Conflicts {
+			fmt.Fprintf(c.Stdout(), "file: %s\n", cf.Path)
+			fmt.Fprintf(c.Stdout(), "conflict type: %s\n", cf.Action)
+			if cf.OursPath != "" {
+				fmt.Fprintf(c.Stdout(), "our file was renamed to: %s\n", cf.OursPath)
+			}
+			if cf.IncomingTemplatePath != "" {
+				fmt.Fprintf(c.Stdout(), "incoming file: %s\n", cf.IncomingTemplatePath)
+			}
+			fmt.Fprintf(c.Stdout(), "--\n")
 		}
-		fmt.Fprintf(c.Stdout(), "--\n")
+		return &common.ExitCodeError{Code: 2}
+	case upgrade.PatchReversalConflict:
+		// TODO(upgrade): include instructions for resolving these conflicts
+		return &common.ExitCodeError{Code: 3}
```

Comment:
Fancy. Just checked even git doesn't differentiate these exit codes :)

I wonder if this should be specific error types. Exit codes are for CLI but not super library friendly. Non blocking though.

Reply:
This file is just for the CLI command, not the library function. The library function for Upgrade() returns a status enum, which kinda does what you suggest.